source/uri.c

```diff
@@ -0,0 +1,380 @@
+void aws_uri_clean_up(struct aws_uri *uri) {
+    if (uri->uri_str.allocator) {
+        aws_byte_buf_clean_up(&uri->uri_str);
+    }
+    AWS_ZERO_STRUCT(*uri);
+}
+
+const struct aws_byte_cursor *aws_uri_scheme(const struct aws_uri *uri) {
+    return &uri->scheme;
+}
+
+const struct aws_byte_cursor *aws_uri_authority(const struct aws_uri *uri) {
+    return &uri->authority;
+}
+
+const struct aws_byte_cursor *aws_uri_path(const struct aws_uri *uri) {
+    return &uri->path;
+}
+
+const struct aws_byte_cursor *aws_uri_query_string(const struct aws_uri *uri) {
+    return &uri->query_string;
+}
+
+const struct aws_byte_cursor *aws_uri_request_uri(const struct aws_uri *uri) {
+    return &uri->request_uri;
+}
+
+const struct aws_byte_cursor *aws_uri_host_name(const struct aws_uri *uri) {
+    return &uri->host_name;
+}
+
+uint16_t aws_uri_port(const struct aws_uri *uri) {
+    return uri->port;
+}
+
+int aws_uri_query_string_params(const struct aws_uri *uri, struct aws_array_list *params) {
+    if (uri->query_string.len == 0) {
+        return AWS_OP_SUCCESS;
+    }
+
+    struct aws_array_list key_val_array;
+    if (aws_array_list_init_dynamic(&key_val_array, uri->allocator, 8, sizeof(struct aws_byte_cursor))) {
+        return AWS_OP_ERR;
+    }
+
+    if (aws_byte_cursor_split_on_char(&uri->query_string, '&', &key_val_array)) {
+        goto error;
+    }
+
+    for (size_t i = 0; i < aws_array_list_length(&key_val_array); ++i) {
+        struct aws_byte_cursor *key_val = NULL;
+        aws_array_list_get_at_ptr(&key_val_array, (void *)&key_val, i);
+
+        uint8_t *delim = memchr(key_val->ptr, '=', key_val->len);
+
+        if (delim) {
+            struct aws_uri_params param_value = {
+                .key =
+                    {
+                        .ptr = key_val->ptr,
+                        .len = delim - key_val->ptr,
+                    },
+                .value =
+                    {
+                        .ptr = delim + 1,
+                        .len = key_val->len - (delim - key_val->ptr) - 1,
+                    },
+            };
+
+            if (aws_array_list_push_back(params, &param_value)) {
+                goto error;
+            }
+        }
+    }
+    aws_array_list_clean_up(&key_val_array);
+    return AWS_OP_SUCCESS;
+error:
+    aws_array_list_clean_up(&key_val_array);
+    aws_array_list_clear(params);
+    return AWS_OP_ERR;
+}
+
+static void s_parse_scheme(struct uri_parser *parser, struct aws_byte_cursor *str) {
+    uint8_t *location_of_colon = memchr(str->ptr, ':', str->len);
+
+    if (!location_of_colon) {
+        parser->state = ON_AUTHORITY;
+        return;
+    }
+
+    /* make sure we didn't just pick up the port by mistake */
+    if ((size_t)(location_of_colon - str->ptr) < str->len && *(location_of_colon + 1) != '/') {
+        parser->state = ON_AUTHORITY;
+        return;
+    }
+
+    parser->uri->scheme.ptr = str->ptr, parser->uri->scheme.len = location_of_colon - str->ptr;
```

Comment:
NONONONONONONONO

Why you use the comma operator? Make this 2 lines please.

Reply:
It was an accident left over from a declared initializer removal.... chill